Add write approval callbacks for Failsafe values of CS-LPC usecase

sthelen-enqs opened the pull request:
This is a work-in-progress proposal for adding write approval callbacks for the failsafe values of the CS-LPC usecase.

Code changes should be functional, but are not yet final. Cleanup and thorough testing may still be required.

Fixes/updates from our side should come when I can find the time to do so. If someone wants to use/update this in the meantime, feel free.

Comments/fixes etc are welcome.

LucaRoth-EnQS committed:
First implementation of approval callback mechanism for failsafe limit writes.

LucaRoth-EnQS committed:
Added FailsafeDurationMinimum to approval mechanism for device config feature.

Made implementation of approval mechansim more general towards the device configuration feature.

LucaRoth-EnQS committed:
refined some comments and a variable name

LucaRoth-EnQS committed:
Provide a more uniform interface for retrieving pending writes

LucaRoth-EnQS committed:
Add config's full description to the PendingDeviceConfiguration return

This allows to more easily check for what a new value is trying to be set and what kind of value is being set (i.e. a duration, a bool, string etc.)

DerAndereAndi commented:
Thanks for starting the work on this. I do now think that the write approval callback API should be used for all writes on any server feature.

sthelen-enqs committed:
Reformat files and fix some inconsistencies

sthelen-enqs commented:
I was waiting for some final qualification tests to ensure that a "spec-compliant" implementation can be achieved for LPC/LPP with this PR. The final test is still outstanding due to delays with the qualification, but I'm reasonably confident that no further changes are necessary to this PR to fully pass the qualification.

DerAndereAndi commented:
@sthelen-enqs I analyzed the PR with Claude and iterated a bit over it. Here is its summary:

Based on my analysis, the PR implementation is well-designed and follows the established patterns correctly. Here are my recommendations:

  1. Fix the Logging Issue

  In deviceConfigurationWriteCB, fix the format string:

```
  // Current (line 228)
  logging.Log().Debug("LPC deviceConfigurationWriteCB: no device configuration for KeyID %d found")

  // Should be:
  logging.Log().Debug("LPC deviceConfigurationWriteCB: no device configuration for KeyID found: ", *deviceKeyValueData.KeyId)
```

  2. Add Documentation

  Add a comment in the interface or README explaining the approval pattern:

  // WriteApprovalRequired event is fired when any server feature receives a write request
  // that requires approval. The application should:
  // 1. Check PendingConsumptionLimits() for LoadControl writes
  // 2. Check PendingDeviceConfigurations() for DeviceConfiguration writes
  // 3. Call the appropriate ApproveOrDeny* method with the msgCounter

  3. Consider Adding Helper Methods (Optional)

  For better developer experience, consider adding helper methods to check if specific configurations are pending:

```
  // Check if a specific configuration key is pending approval
  func (e *LPC) IsPendingDeviceConfiguration(msgCounter model.MsgCounterType, keyName model.DeviceConfigurationKeyNameType) bool {
      configs := e.PendingDeviceConfigurations()
      if pendingConfigs, exists := configs[msgCounter]; exists {
          for _, config := range pendingConfigs {
              if config.Description != nil && config.Description.KeyName != nil &&
                 *config.Description.KeyName == keyName {
                  return true
              }
          }
      }
      return false
  }
```

  4. Test Coverage

  Ensure tests cover:
  - Approval flow for both parameters
  - Auto-approval of unrelated DeviceConfiguration writes
  - Handling multiple configuration changes in one message
  - Thread safety of concurrent approvals
  - Timeout behavior (inherited from spine-go)

  Summary

  The PR implementation is functionally complete and well-designed. It correctly:
  - Follows the established LoadControl pattern
  - Provides selective filtering for relevant parameters
  - Maintains thread safety
  - Gives applications full control over approval decisions
  - Auto-approves unrelated writes to avoid blocking other use cases

  The only required change is the logging fix. The other suggestions are optional improvements for developer experience and documentation.

sthelen-enqs committed:
Update WriteApprovalRequired event documentation with new methods

sthelen-enqs committed:
Fix logging to include the KeyID for which no configuration exists

sthelen-enqs commented:
1. The logging issue should be fixed now
2. I updated the documentation in events.go for LPC/LPP with the new methods as well.
3. I don't think point 3 is particularly relevant since you need to check every pending device configuration (and then switch on the type) in the handler, I wouldn't recommend selectively handling device configurations.
4. I'll double-check the test coverage for this

DerAndereAndi commented on usecases/cs/lpc/usecase.go:
```diff
@@ -172,6 +177,80 @@ func (e *LPC) loadControlWriteCB(msg *spineapi.Message) {
+		}
+
+		// Only ask for write approval if at least one of the configurations we care about is trying to be set
+		if _, exists := configsToApprove[*description.KeyName]; exists {
```
Doesn't this panic when KeyName is nil?

DerAndereAndi commented on usecases/cs/lpp/usecase.go:
```diff
@@ -172,6 +177,80 @@ func (e *LPP) loadControlWriteCB(msg *spineapi.Message) {
+		}
+
+		// Only ask for write approval if at least one of the configurations we care about is trying to be set
+		if _, exists := configsToApprove[*description.KeyName]; exists {
```
Doesn't this panic when KeyName is nil?

DerAndereAndi commented on examples/hems/main.go:
```diff
@@ -152,12 +152,21 @@ func (h *hems) OnLPCEvent(ski string, device spineapi.DeviceRemoteInterface, ent
+		for msgCounter, configs := range pendingDeviceConfigWrites {
+			fmt.Printf("Approving LPC device config write with msgCounter %d for features: ", msgCounter)
+			for _, config := range configs {
+				fmt.Printf("%s ", *config.Description.KeyName)
```
What if Description or Keyname are nil?

DerAndereAndi commented on usecases/cs/lpp/public.go:
```diff
@@ -267,6 +267,60 @@ func (e *LPP) SetFailsafeDurationMinimum(duration time.Duration, changeable bool
 	return dc.UpdateKeyValueDataForFilter(data, nil, filter)
 }
 
+// return the currently pending incoming failsafe consumption limit writes
```
`consumption` -> `production`

DerAndereAndi commented on usecases/cs/lpc/usecase.go:
```diff
@@ -172,6 +177,80 @@ func (e *LPC) loadControlWriteCB(msg *spineapi.Message) {
+
+	data := msg.Cmd.DeviceConfigurationKeyValueListData
+
+	if data == nil || data.DeviceConfigurationKeyValueData == nil || len(data.DeviceConfigurationKeyValueData) == 0 {
```
data can't be nil as it is assigned to `msg.Cmd.DeviceConfigurationKeyValueListData` which in line 201 is checked not to be nil. So not wrong, but will never be the case anyway

DerAndereAndi commented on usecases/api/types.go:
```diff
@@ -166,3 +166,9 @@ type DurationSlotValue struct {
+type PendingDeviceConfiguration struct {
+	Description       *model.DeviceConfigurationKeyValueDescriptionDataType `json:"description,omitempty"`
+	Value             *model.DeviceConfigurationKeyValueValueType           `json:"value,omitempty"`
+	IsValueChangeable *bool                                                 `json:"isValueChangeable,omitempty" eebus:"writecheck"`
```
Why did you add the `writecheck` tag attribute here?

DerAndereAndi reviewed (changes requested):
In addition to my comments, we are missing tests for the new code.

Additionally `WriteApprovalRequired` message is now used for limit write approvals and failsafe and the application doesn't know which and both could also happen around the same time. In this case I would propose introduce a new message so the application can handle it more easily.

The newmethods in `usecase.go` and `public.go` are nearly identical. We could move that code into shared functions in the `internal` package.

LucaRoth-EnQS committed:
Refactored device config write approval functionality that both LPC and LPP use into shared internal functions. Also added tests.

sthelen-enqs committed:
Clarify comments describing events

sthelen-enqs committed:
Simplify and clean up logging

DerAndereAndi commented on usecases/internal/deviceconfiguration.go:
```diff
@@ -0,0 +1,91 @@
+		return result
+	}
+
+	for msgCounter, msg := range pendingDeviceConfigs {
+		data := msg.Cmd.DeviceConfigurationKeyValueListData
+		for _, configKeyValueData := range data.DeviceConfigurationKeyValueData {
+			description, err := dc.GetKeyValueDescriptionFoKeyId(*configKeyValueData.KeyId)
+			if err != nil {
+				continue
+			}
+
+			pendingConfigData := ucapi.PendingDeviceConfiguration{
+				Description:       description,
+				Value:             configKeyValueData.Value,
+				IsValueChangeable: configKeyValueData.IsValueChangeable,
+			}
+
+			if _, exists := result[msgCounter]; !exists {
+				result[msgCounter] = []ucapi.PendingDeviceConfiguration{pendingConfigData}
+			} else {
+				result[msgCounter] = append(result[msgCounter], pendingConfigData)
+			}
+		}
+	}
```
How about changing this to:

```go
for msgCounter, msg := range pendingDeviceConfigs {
    if msg.Cmd.DeviceConfigurationKeyValueListData == nil {
        continue
    }
    for _, configKeyValueData := range msg.Cmd.DeviceConfigurationKeyValueListData.DeviceConfigurationKeyValueData {
        if configKeyValueData.KeyId == nil {
            continue
        }
        description, err := dc.GetKeyValueDescriptionFoKeyId(*configKeyValueData.KeyId)
        if err != nil || description == nil || description.KeyName == nil {
            continue
        }
        result[msgCounter] = append(result[msgCounter], ucapi.PendingDeviceConfiguration{
            Description:       description,
            Value:             configKeyValueData.Value,
            IsValueChangeable: configKeyValueData.IsValueChangeable,
        })
    }
}
```

This way the example code does not need to check for `if config.Description.Keyname != nil` and also `config.Description` being `nil` is also avoided.

So the hems example would only have this:

```go
for _, config := range configs {
    fmt.Printf("%s ", string(*config.Description.KeyName))
}
```

DerAndereAndi commented on usecases/cs/lpc/usecase.go:
```diff
@@ -175,6 +179,51 @@ func (e *LPC) loadControlWriteCB(msg *spineapi.Message) {
+		return
+	}
+
+	if approvalRequired {
+		e.pendingDeviceConfigMux.Lock()
+		if _, exists := e.pendingDeviceConfigs[*msg.RequestHeader.MsgCounter]; !exists {
+			e.pendingDeviceConfigs[*msg.RequestHeader.MsgCounter] = msg
+			e.pendingDeviceConfigMux.Unlock()
+			e.EventCB(msg.DeviceRemote.Ski(), msg.DeviceRemote, msg.EntityRemote, ConfigurationWriteApprovalRequired)
+			return
+		}
+		e.pendingDeviceConfigMux.Unlock()
+	}
+
+	// If neither a failsafe duration nor a failsafe limit were set this message does not pertain to this use case so we accept
+	go e.approveOrDenyDeviceConfiguration(msg, true, "")
+}
```
`if approvalRequired==true` and the msgCounter is already in `pendingDeviceConfigs`, the code unlocks and then auto-approves the duplicate — approving a message whose original is still pending and leaving the original entry orphaned. This issue is pre-existing, but worth checking if we should improve this maybe?

DerAndereAndi commented on usecases/cs/lpc/public.go:
```diff
@@ -270,6 +271,31 @@ func (e *LPC) SetFailsafeDurationMinimum(duration time.Duration, changeable bool
 	return dc.UpdateKeyValueDataForFilter(data, nil, filter)
 }
 
+// return the currently pending incoming failsafe consumption limit writes
```
Shouldn't this rather be:

```
// return the currently pending incoming device configuration writes
```

DerAndereAndi commented on usecases/cs/lpp/public.go:
```diff
@@ -272,6 +273,32 @@ func (e *LPP) SetFailsafeDurationMinimum(duration time.Duration, changeable bool
 	return dc.UpdateKeyValueDataForFilter(data, nil, filter)
 }
 
+// return the currently pending incoming failsafe production limit writes
```
Shouldn't this rather be:

```
// return the currently pending incoming device configuration writes
```

DerAndereAndi reviewed (changes requested):
An additional note regarding LPC and LPP `PendingDeviceConfigurations()`: if LPC and LPP are on the same entity, both need to be approved, even though it is the same data point! The application has to be aware of this and we may want to think of adding tests for this?
We also should to re-check if spine-go handles this correctly.

TomLucaRoth committed:
Implemented suggestions from PR:
- Add early return to deviceConfigurationWriteCB for both LPC and LPP so possible duplicate messages do not just result in accepting the write
- Change control flow of GroupPendingDeviceConfigurations slightly, but most importantly, verify that description and description.KeyName are non-nil
- Updated PendingDeviceConfiguration struct, to encode the fact that Description and KeyName are now guaranteed non nil (non-nil guaranteed by the type not just by convention)

TomLucaRoth committed:
Updated and extended tests to reflect/cover changes made to GroupPendingDeviceConfigurations

TomLucaRoth committed:
Fixed comments on PendingDeviceConfiguration methods

TomLucaRoth committed:
Added integration test to verify behavior of writing to shared FailsafeDurationMinium datapoint  when an entity has both LPC and LPC use cases
